master/internal/resourcemanagers/fitting_methods.go

```diff
@@ -12,21 +14,50 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
 	return req.Label == agent.label
 }
 
+func maxZeroSlotContainersSatisfied(req *AllocateRequest, agent *agentState) bool {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == 0 {
+			return false
+		}
+		return agent.numZeroSlotContainers() < agent.maxZeroSlotContainers
+	}
+	return true
+}
+
+func agentSlotUnusedSatisfied(_ *AllocateRequest, agent *agentState) bool {
+	return agent.numUsedSlots() == 0
+}
+
 // Soft Constraints
 
 // BestFit returns a float affinity score between 0 and 1 for the affinity between the task and
 // the agent. This method attempts to allocate tasks to the agent that is both most utilized and
 // offers the fewest slots. This method should be used when the cluster is dominated by multi-slot
 // applications.
-func BestFit(_ *AllocateRequest, agent *agentState) float64 {
-	return 1.0 / (1.0 + float64(agent.numEmptySlots()))
+func BestFit(req *AllocateRequest, agent *agentState) float64 {
+	switch {
+	case agent.numUsedSlots() != 0 || req.SlotsNeeded != 0:
+		return 1.0 / (1.0 + float64(agent.numEmptySlots()))
+	case agent.maxZeroSlotContainers == 0:
```

Comment:
I'm confused by this line - if we aren't allowing any zero slot containers on this agent, the affinity should be zero right?

Reply:
Yeah this brings up a good point where we are defining `maxZeroSlotContainers` 0 as the default, where a user may want to actually set it to zero. Seems like we should change it to an `*int`.